fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexer.java

```diff
@@ -994,42 +1002,56 @@ public void rebuildIndex(@Nonnull FDBRecordStore store) {
-     *         there is another runner actively working on the same index.
-     *     </li>
-     *     <li>
-     *         It checks and updates index states and clear index data respecting the {@link IndexStatePrecondition}. So there is
-     *         no need to change the index states or clear existing index data before invoking this method.
-     *     </li>
-     * </ul>
-     * @return a future that will be ready when the build has completed.
+     * Builds an index across multiple transactions.
+     * <p>
+     * If it is set to use synchronized sessions, it stops with {@link com.apple.foundationdb.synchronizedsession.SynchronizedSessionLockedException}
+     * when there is another runner actively working on the same index. It first checks and updates index states and
+     * clear index data respecting the {@link IndexStatePrecondition} being set. It then builds the index across
+     * multiple transactions honoring the rate-limiting parameters set in the constructor of this class. It also retries
+     * any retriable errors that it encounters while it runs the build. At the end, it marks the index readable in the
+     * store.
+     * </p>
+     * <p>
+     * One may consider to set the index state precondition to {@link IndexStatePrecondition#ERROR_IF_DISABLED_CONTINUE_IF_WRITE_ONLY}
+     * and {@link OnlineIndexer.Builder#setUseSynchronizedSession(boolean)}} to {@code false}, which makes the indexer
+     * follow the same behavior as before. But it is not recommended.
+     * </p>
+     * @return a future that will be ready when the build has completed
      * @throws com.apple.foundationdb.synchronizedsession.SynchronizedSessionLockedException the build is stopped
      * because there may be another build running actively on this index.
      */
-    // TODO: It should support range build (for building index in parallel) as well.
     @Nonnull
-    public CompletableFuture<Void> safelyBuildIndexAsync() {
-        return safelyBuildIndexAsync(true);
+    public CompletableFuture<Void> buildIndexAsync() {
+        return buildIndexAsync(true);
     }
 
     @VisibleForTesting
     @Nonnull
-    CompletableFuture<Void> safelyBuildIndexAsync(boolean markReadable) {
+    CompletableFuture<Void> buildIndexAsync(boolean markReadable) {
+        if (!useSynchronizedSession) {
+            synchronizedSessionRunner = null;
+            return handleStateAndDoBuildIndexAsync(markReadable);
+        }
         return runner
-                .runAsync(context -> openRecordStore(context).thenApply(store -> store.indexBuildLockSubspace(index)))
-                .thenCompose(lockSubspace -> runner.startSynchronizedSessionAsync(lockSubspace, DEFAULT_LEASE_LENGTH_MILLIS))
+                .runAsync(context -> openRecordStore(context).thenApply(store -> indexBuildLockSubspace(store, index)))
+                .thenCompose(lockSubspace -> runner.startSynchronizedSessionAsync(lockSubspace, leaseLengthMills))
```

Comment:
Given that there is now only the one `buildIndexAsync` method (and not a separate `safelyBuildIndex` method), is it worth it to integrate that into all of the range-based methods?

I'm not quite sure what the semantics should be for the ones that take a transaction, but I think for the ones that don't, it would mean I think abstracting these two lines here and the `IndexStatePrecondition` part (unless we think that that part doesn't make sense for those parts of the API).

If we don't update those in this PR, it might make sense to mention that limitation in the Javadoc somewhere and mention that those are expected to change.

This is separate, to be clear, from making it so that one can re-use the synchronized session ID to join a session, which is kind of the use case where one wants to use the range-based APIs, but I think there is a reasonable enough semantics even without that (though I could be wrong).

Reply:
I don't think one can support building a single index from multiple machines in parallel with a single `buildIndex` method (or at least not without it making RPCs internally to other machines, but then one needs a way to push down endpoints for those other machines and/or introduce an RPC-sending mechanism to the Record Layer...which seems like a lot of work).

I think that these are reasonable enough semantics. Is `buildRange` called internally? In theory, it could be made to throw an error if it is called with an existing synchronized session if you're worried about it. (Having it effectively have `ERROR_IF_DISABLED_CONTINUE_IF_WRITE_ONLY` semantics seems fine given, as you say that it doesn't totally make sense to just build a range in isolation.) At some point, we'll want to have a `buildRange` that takes either a synchronized session or a synchronized session ID or something so that one can join an existing build just to build a range (for parallelized builds), but maybe that's fine for another day. I could also see that being a parameter of the builder itself, or something. Hmmm.

Reply:
`buildRange` *is* called internally - In fact, `buildIndex` calls it to do the real work.

Created a issue for supporting joining a session: https://github.com/FoundationDB/fdb-record-layer/issues/764

Reply:
Right, and the behavior of `buildRange` for internal use-cases kind of needs to be as is. But one could add a `buildRangeInternal` that is used internally and then `buildRange` could throw an error if there were a sycnhronized session or something. I'm not actually sure that's a good idea, but it would be one way to prevent someone inadvertently piggy backing on the same synchronized session.

Reply:
What about enforcing that one `OnlineIndexer` can either `buildRange` or `buildIndex` only once? If the users have to do more than one thing on one indexer, they should use `OnlineIndexer.Builder.build()` to create multiple `OnlineIndexer` instances. This way we can keep the semantic simple especially when it comes to #764. I haven't think of any case where we have to do more than one `buildRange`/`buildIndex` in one `OnlineIndexer`.

Reply:
Hm, interesting. I think you're right that there isn't really a good use case for reusing the same indexer to call `buildIndex` multiple times. With `buildRange`, I could see wanting to use the same object to build disjoint ranges, for example, but maybe not.

If we wanted to make this change (which I suppose we could still do), then we'd probably need to bump the minor version.